File: schemas/podcast-platform/convex/shows.ts
// shows: podcast shows with metadata, feed details, and audience metrics.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const shows = defineTable({
  ownerId: v.id("users"),
  networkId: v.optional(v.id("networks")),
  title: v.string(),
  slug: v.string(),
  description: v.string(),
  htmlDescription: v.optional(v.string()),
  author: v.string(),
  language: v.string(),
  showType: v.union(v.literal("episodic"), v.literal("serial")),
  explicit: v.boolean(),
  artworkFileId: v.optional(v.id("files")),
  bannerFileId: v.optional(v.id("files")),
  feedUrl: v.optional(v.string()),
  website: v.optional(v.string()),
  copyright: v.optional(v.string()),
  ownerName: v.optional(v.string()),
  ownerEmail: v.optional(v.string()),
  podcastGuid: v.optional(v.string()),
  medium: v.union(
    v.literal("podcast"),
    v.literal("music"),
    v.literal("video"),
    v.literal("audiobook"),
    v.literal("newsletter")
  ),
  isComplete: v.boolean(),
  episodeCount: v.number(),
  subscriberCount: v.number(),
  updatedAt: v.number(),
})
  .index("by_slug", ["slug"])
  .index("by_owner_id", ["ownerId"])
  .index("by_network_id", ["networkId"])
  .index("by_language_and_show_type", ["language", "showType"])
  .index("by_subscriber_count", ["subscriberCount"]);

File: schemas/podcast-platform/spacetimedb/shows.rs
// shows: Podcast or media shows with metadata and feed information.
// See README.md for full design rationale.

use spacetimedb::SpacetimeType;
use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum ShowType {
    Episodic, // type: String
    Serial,
}

#[derive(SpacetimeType, Clone)]
pub enum ShowMedium {
    Podcast, // type: String
    Music,
    Video,
    Audiobook,
    Newsletter,
}

#[spacetimedb::table(name = shows, public)]
pub struct Show {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub owner_id: String, // UUID — FK → users.id (cascade delete)
    #[index(btree)]
    pub network_id: Option<String>, // UUID — FK → networks.id (set null on delete)
    pub title: String,
    #[unique]
    pub slug: String,
    pub description: String,
    pub html_description: Option<String>,
    pub author: String,
    pub language: String,
    pub show_type: ShowType,
    pub explicit: bool,
    pub artwork_file_id: Option<String>, // UUID — FK → files.id (set null on delete)
    pub banner_file_id: Option<String>, // UUID — FK → files.id (set null on delete)
    pub feed_url: Option<String>,
    pub website: Option<String>,
    pub copyright: Option<String>,
    pub owner_name: Option<String>,
    pub owner_email: Option<String>,
    pub podcast_guid: Option<String>,
    pub medium: ShowMedium,
    pub is_complete: bool,
    pub episode_count: i32,
    #[index(btree)]
    pub subscriber_count: i32,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
// Composite index: (language, show_type) — not supported, enforce in reducer logic
